static/js/publisher/form/media.js

```diff
@@ -0,0 +1,270 @@
+  constructor(props) {
+    super(props);
+
+    this.markForDeletion = this.markForDeletion.bind(this);
+    this.addImage = this.addImage.bind(this);
+    this.mediaChange = this.mediaChange.bind(this);
+    this.keyboardEvent = this.keyboardEvent.bind(this);
+    this.focused = this.focused.bind(this);
+    this.blurred = this.blurred.bind(this);
+
+    this.state = {
+      mediaData: props.mediaData,
+      focused: null
+    };
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    this.props.updateState(prevState.mediaData);
+
+    if (
+      this.state.focused === "new" &&
+      document.activeElement !== this.blankItem
+    ) {
+      this.blankItem.focus();
+    }
+  }
+
+  markForDeletion(key) {
+    let index = -1;
+    let newFocus = this.state.focused;
+
+    const newMediaData = this.state.mediaData.map((item, i) => {
+      if (item.url === key) {
+        item.status = "delete";
+        index = i;
+      }
+      return item;
+    });
+
+    if (this.state.focused && index > -1) {
+      const available = newMediaData.filter(item => item.status !== "delete");
+      if (available[0]) {
+        newFocus = available[0].url;
+      } else {
+        newFocus = "new";
+      }
+    }
+
+    this.setState({
+      mediaData: newMediaData,
+      focused: newFocus
+    });
+  }
+
+  mediaChange(input) {
+    const newMediaData = this.state.mediaData;
+    for (let i = 0; i < input.files.length; i++) {
+      const file = input.files[i];
+      newMediaData.push({
+        file,
+        url: URL.createObjectURL(file),
+        name: file.name,
+        type: "screenshot",
+        status: "new"
+      });
+    }
+
+    this.setState({
+      mediaData: newMediaData
+    });
+  }
+
+  addImage() {
+    const input = document.createElement("input");
+    input.type = "file";
+    input.accept = "image/x-png,image/gif,image/jpeg";
+    input.name = "screenshots";
+    input.hidden = "hidden";
+    input.addEventListener("change", () => {
+      this.mediaChange(input);
+    });
+
+    this.holder.appendChild(input);
+    input.click();
+  }
+
+  keyboardEvent(e) {
+    if (e.key === "Enter") {
+      this.addImage();
+    }
+  }
+
+  focused(url) {
+    this.setState({
+      focused: url
+    });
+  }
+
+  blurred(url) {
+    if (this.state.focused === url) {
+      this.setState({
+        focused: null
+      });
+    }
+  }
+
+  renderOverLimit() {
+    if (
+      this.state.mediaData.filter(item => item.status !== "delete").length >
+      this.props.mediaLimit
+    ) {
+      return (
+        <div className="p-notification--caution">
+          <p className="p-notification__response">
+            You have over 5 images uploaded. Not all image will be visible to
+            users.
+          </p>
+        </div>
+      );
+    }
+    return false;
+  }
+
+  renderBlankMedia(numberOfBlank) {
+    const blankMedia = [];
+    for (let i = 0; i < numberOfBlank; i++) {
+      const classes = mediaClasses.slice(0);
```

Comment:
That `slice` confused me, didn't realise it's just cloning the array.
We have transpilers in place, maybe that would be clearer:
```suggestion
      const classes = [...mediaClasses];
```